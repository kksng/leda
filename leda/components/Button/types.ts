import React from 'react';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { Form, FormsObject, ValidationButtonProps } from '../Validation/types';

export interface SubmitEvent extends React.MouseEvent<HTMLButtonElement> {
  form?: FormsObject,
  forms?: Form[],
}

export type ValidationFailEvent = React.MouseEvent<HTMLButtonElement> & { invalidForms: Form[] };

export interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement>, ValidationButtonProps {
  /** Название формы */
  form?: string | string[],
  /** Отключение кнопки */
  isDisabled?: boolean,
  /** Показывать лоадер */
  isLoading?: boolean,
  /** Обработчик клика */
  onClick?: (event: SubmitEvent) => void,
  /** Применяется к кнопке. Обработчик, который срабатывает, если есть невалидные поля,
   * если срабатывает onValidationFail, то onClick на кнопке не сработает */
  onValidationFail?: (ev: ValidationFailEvent) => void,
  /** Реф */
  ref?: React.Ref<ButtonRefCurrent>,
  /** Задержкаа перед скроллм к невалидным полям */
  scrollDelay?: number,
  /** Дополнительный сдвиг в px при скролинге к кнопке */
  scrollOffset?: number,
  /** Если передано false, компонент не будет рендериться */
  shouldRender?: boolean,
  /** Прокрутка к невалидным полям при нажатии на кнопку */
  shouldScrollToInvalidFields?: boolean,
  /** Тема компонета */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.button],
  /** Тип кнопки */
  type?: React.ButtonHTMLAttributes<{}>['type'],
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface ButtonRefCurrent {
  wrapper: HTMLButtonElement | null,
}
