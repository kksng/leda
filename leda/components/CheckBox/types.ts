import * as React from 'react';
import { CustomRender } from '../../commonTypes';
import { PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { COMPONENTS_NAMESPACES } from '../../constants';

export interface ChangeEvent extends React.ChangeEvent<HTMLInputElement> {
  component: {
    value: boolean,
    name?: string,
  },
}

export interface CheckBoxProps {
  /** Всё, что обёрнуто в L.Checkbox, попадает в label */
  children?: React.ReactNode,
  /** Значение по-умолчанию, если не передано - false */
  defaultValue?: boolean,
  /** Id для чекбокса */
  id?: string,
  /** Кастомизация инпута, непосредственно инпут - невидим, но данный рендер позволяет добавить атрибуты в тег <input> */
  inputRender?: CustomRender<CheckBoxProps, {}, React.InputHTMLAttributes<HTMLInputElement>>,
  /** Отключение чекбокса */
  isDisabled?: boolean,
  
  /** Флаг частичного выбора */
  isSemi?: boolean,

  /** Кастомизация элемента label */
  labelRender?: CustomRender<CheckBoxProps, {}, React.LabelHTMLAttributes<any>>,
  /** Имя компонента */
  name?: string,
  /** Обработчик изменения состояния элементов */
  onChange?: (event: ChangeEvent) => void,
  /** Реф */
  ref?: React.Ref<CheckBoxRefCurrent>,
  /** Тема компонента */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.checkBox],
  /** Значение чекбокса */
  value?: boolean,
  /** Кастомизация враппера */
  wrapperRender?: CustomRender<CheckBoxProps, {}, Partial<CheckBoxProps>>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface CheckBoxRefCurrent {
  wrapper: HTMLElement | null,
  input: HTMLInputElement | null,
  label: HTMLLabelElement | null,
}
