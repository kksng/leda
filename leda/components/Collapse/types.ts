import * as React from 'react';
import { CustomRender, CustomEventHandler } from '../../commonTypes';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { GlobalDefaultTheme, PartialGlobalDefaultTheme } from '../../utils/useTheme';

export interface SelectEvent {
  component: {
    value: string | string[] | null,
  },
}

export interface BodyClickCustomEvent {
  component: {
    value: string, // panelKey
  },
}


export interface CollapseProps {
  /** Признак активной панели. Сюда передается значение panelKey активной панели. Когда панель(панели) закрыта, значение равно null */
  activePanelKey?: string | string[] | null,
  /** Дочерние элементы */
  children: React.ReactNode,
  /** Признак возможности открытия нескольких панелей для неконтролируемого режима */
  isAccordion?: boolean,
  /** Обработчик клика на панель */
  onSelect?: CustomEventHandler<SelectEvent>,
  /** Тема */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.collapse],
}

export interface HeadingProps {
  /** Дочерние элементы */
  children: React.ReactNode,
  /** Иконка возле заголовка */
  iconRender?: CustomRender<HeadingProps, { }, IconProps>,
  /** Обработчик клика */
  onClick?: React.MouseEventHandler<HTMLDivElement>,
  /** Реф */
  ref?: React.Ref<HeadingRefCurrent>,
  /** Компонент-обертка в который будет помещен Collapse.Heading. Передавать в виде <Wrapper props />. По умолчанию - <Div /> */
  wrapperRender?: CustomRender<HeadingProps, { }, HeadingWrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface BodyProps {
  /** Дочерние элементы */
  children?: React.ReactNode,
  /** Состояние загрузки данных */
  isLoading?: boolean,
  /** Функция обратного вызова на закрытие коллапса */
  onClose?: CustomEventHandler<BodyClickCustomEvent>,
  /** Функция обратного вызова на закрытие коллапса по клику */
  onCloseByClick?: CustomEventHandler<BodyClickCustomEvent>,
  /** Функция обратного вызова на открытие коллапса */
  onOpen?: CustomEventHandler<BodyClickCustomEvent>,
  /** Реф */
  ref?: React.Ref<BodyRefCurrent>,
  /** Описание открытия/закрытия коллапса. Описывается как CSS Transition. См. https://developer.mozilla.org/ru/docs/Web/CSS/transition. По умолчанию height 250ms cubic-bezier(.4, 0, .2, 1) */
  transition?: string,
  /** Компонент-обертка в который будет помещен Collapse.Body. Передавать в виде <Wrapper props />. По умолчанию - <Div /> */
  wrapperRender?: CustomRender<BodyProps, { }, BodyWrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface PanelProps {
  /** Дочерние элементы */
  children: React.ReactNode,
  /** Отключение панели */
  isDisabled?: boolean,
  /** Имя панели */
  name?: string,
  /** Идентификатор события панели, данное значение передается в activeKey компонента Collapse */
  panelKey: string,
  /** Компонент-обертка для панели, например Div. Передавать в виде <Wrapper props /> */
  wrapperRender?: CustomRender<PanelProps, { isClicked: boolean }, PanelWrapperProps>,
  /** Классы переданные через _ */
  [x: string]: unknown,
}

export interface CollapseContextType {
  activePanelKey: string | string[] | null,
  onSelect: SelectHandler,
  theme: GlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.collapse],
}

export interface CollapsePanelContextType extends CollapseContextType {
  panelKey: string,
  isClicked: boolean,
  isDisabled?: boolean,
  name?: string,
  isExpanded: boolean,
  onHeadingClick: CollapsePanelHandlers['onHeadingClick'],
  onBodyRest: CollapsePanelHandlers['onBodyRest'],
}

export interface HeadingWrapperProps {
  children?: React.ReactNode,
  className?: string,
  onClick: React.MouseEventHandler,
}

export interface IconProps {
  className?: string,
  isExpanded: boolean,
}

export interface BodyWrapperProps {
  className?: string,
  children?: React.ReactNode,
}

export interface PanelWrapperProps {
  children?: React.ReactNode,
  className?: string,
}

export type KeyState = string | string[] | null;

export interface SelectHandler {
  (ev: BodyClickCustomEvent): void,
}

export type SetClicked = React.Dispatch<React.SetStateAction<boolean>>;

export interface CollapsePanelHandlers {
  onHeadingClick: () => void,
  onBodyRest: () => void,
}

export type CollapseComponent = React.FC<CollapseProps> & { Panel: React.FC<PanelProps>, Body: React.FC<BodyProps>, Heading: React.FC<HeadingProps> };

export interface BodyRefCurrent {
  wrapper: HTMLElement | null,
}

export interface HeadingRefCurrent {
  wrapper: HTMLElement | null,
}
