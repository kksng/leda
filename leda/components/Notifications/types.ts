import * as React from 'react';
import { GlobalDefaultTheme, PartialGlobalDefaultTheme } from '../../utils/useTheme';
import { COMPONENTS_NAMESPACES } from '../../constants';
import { CustomRender } from '../../commonTypes';
import { DivProps } from '../Div';
import { ChangeMethods } from './constants';

export interface Item {
  // текст, можно использовать html
  text: string,
  // по умолчанию 5000 мс, чтобы уведомление не закрывалось по timeout - передайте 0
  delay?: number,
  id: string | number,
  className?: string,
  iconClassName?: string,
  // можно передать любые дополнительные данные
  [x: string]: unknown,
}

// todo: extend ChangeEvent
export interface ChangeEvent {
  component: {
    value: Item[],
    name?: string,
    method: ChangeMethods,
  },
}

export interface NotificationItemProps {
  /** Объект с уведомлением */
  item: Item,
  /** Кастомизация иконки */
  iconRender?: CustomRender<NotificationItemProps, {}, DivProps>,
  /** Кастомизация кнопки внутри оповещения, по-умолчанию не отображается */
  actionButtonRender?: CustomRender<NotificationItemProps, {}, React.PropsWithChildren<{}>>,
  /** Обработчик изменения, срабатывает при клике по крестику или по таймауту */
  onChange: (item: Item, method: ChangeMethods) => void,
  /** Сообщение внутри оповещения. Принимается функция. */
  contentRender?: CustomRender<NotificationItemProps, {}, DivProps>,
  /** Тема компонента */
  theme: GlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.notifications],
}

export interface NotificationsProps {
  /** Сообщение внутри оповещения. Принимается функция. */
  contentRender?: CustomRender<NotificationItemProps, {}, DivProps>,
  /** Кастомизация иконки */
  iconRender?: CustomRender<NotificationItemProps, {}, DivProps>,
  /** Кастомизация кнопки внутри оповещения, по-умолчанию не отображается */
  actionButtonRender?: CustomRender<NotificationItemProps, {}, React.PropsWithChildren<{}>>,
  /** Уведомления, если передано больше, чем maxItems - отображаются в порядке очереди */
  value: Item[],
  /** Максимальное количество оповещений на экране, по-умолчанию 3 */
  maxItems?: number,
  /** Реф */
  ref?: React.Ref<NotificationRefCurrent>,
  /** Обработчик изменения, срабатывает при клике по крестику или по таймауту  */
  onChange: (event: ChangeEvent) => void,
  /** Тема компонента */
  theme?: PartialGlobalDefaultTheme[typeof COMPONENTS_NAMESPACES.notifications],
  /** Классы, переданные через _ */
  [x: string]: unknown,
}

export interface NotificationRefCurrent {
  wrapper: HTMLElement | null,
}
