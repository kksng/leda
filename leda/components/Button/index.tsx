import React from 'react';
import { COMPONENTS_NAMESPACES } from '../../constants';
import {
  getClassNames, bindFunctionalRef, useTheme, useProps,
} from '../../utils';
import { createClickHandler } from './handlers';
import { ButtonProps, ButtonRefCurrent } from './types';

// как настраивать кнопку для валидации ввода: ../Validation/validation.md
export const Button = React.forwardRef((props: ButtonProps, ref: React.Ref<ButtonRefCurrent>): React.ReactElement => {
  const {
    children,
    className,
    isDisabled,
    isLoading,
    form,
    onValidationFail,
    shouldScrollToInvalidFields,
    shouldValidateUnmounted,
    theme: themeProp,
    ...restProps
  } = useProps<ButtonProps>(props);

  const theme = useTheme(themeProp, COMPONENTS_NAMESPACES.button);

  const { disabled, loading, wrapper } = theme;

  const handleClick = createClickHandler(props);

  const combinedClassNames = getClassNames(
    wrapper,
    className,
    { [disabled]: isDisabled },
    { [loading]: isLoading },
  );

  return (
    <button
      type="button"
      {...restProps}
      className={combinedClassNames}
      onClick={handleClick}
      ref={ref && ((component) => bindFunctionalRef(component, ref, component && {
        wrapper: component,
      }))}
    >
      {children}
    </button>
  );
}) as React.FC<ButtonProps>;

Button.displayName = 'Button';
