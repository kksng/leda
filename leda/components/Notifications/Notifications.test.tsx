import React from 'react';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import shortid from 'shortid';
import userEvent from '@testing-library/user-event';
import { Notifications } from './index';
import { ChangeEvent, Item } from './types';
import { Button } from '../Button';

jest.useFakeTimers();

describe('Notifications SNAPSHOTS', () => {
  it('should render basic usage', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 10,
        id: 1,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 15,
        id: 2,
      },
    ];

    const onChange = jest.fn();

    const eventMatcher = expect.objectContaining({
      component: expect.objectContaining({
        value: expect.any(Array),
        method: 'delay',
      }),
    });

    const { container } = render(
      <Notifications
        value={items}
        onChange={onChange}
        maxItems={3}
      />,
    );

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(2);

    expect(container).toMatchSnapshot();

    jest.advanceTimersByTime(100);

    expect(onChange).toHaveBeenCalledTimes(2);

    expect(onChange).toHaveBeenCalledWith(eventMatcher);
  });

  it('should render items < maxItems', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 10,
        id: 1,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 10,
        id: 2,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 10,
        id: 3,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 10,
        id: 4,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 10,
        id: 5,
      },
    ];

    const { container } = render(
      <Notifications
        value={items}
        onChange={jest.fn()}
        maxItems={3}
      />,
    );

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(3);

    expect(container).toMatchSnapshot();
  });

  it('should render action button', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 10,
        id: 1,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 15,
        id: 2,
      },
    ];

    const { container } = render(
      <Notifications
        value={items}
        onChange={jest.fn()}
        actionButtonRender={() => <Button>text</Button>}
        maxItems={3}
      />,
    );

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(2);

    expect(container.querySelectorAll('button')).toHaveLength(2);

    expect(container.querySelector('button')?.textContent).toEqual('text');

    expect(container).toMatchSnapshot();
  });
});

describe('Notifications handlers', () => {
  it('should trigger onChange close', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 10,
        id: 1,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 15,
        id: 2,
      },
    ];

    const eventMatcher = expect.objectContaining({
      component: expect.objectContaining({
        value: expect.any(Array),
        method: 'delay',
      }),
    });

    const Wrapper = () => {
      const [value, setValue] = React.useState<Item[]>(items);

      const handleChange = (ev: ChangeEvent) => {
        expect(ev).toMatchObject(eventMatcher);

        setValue(ev.component.value);
      };

      return (
        <Notifications
          value={value}
          onChange={handleChange}
          maxItems={3}
        />
      );
    };

    const { container } = render(<Wrapper />);

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(2);

    expect(container).toMatchSnapshot();

    act(() => {
      jest.advanceTimersByTime(100);
    });

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(0);

    expect(container).toMatchSnapshot();
  });

  it('should not delete items with delay 0', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 0,
        id: 1,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 15,
        id: 2,
      },
    ];

    const eventMatcher = expect.objectContaining({
      component: expect.objectContaining({
        value: expect.any(Array),
        method: 'delay',
      }),
    });

    const Wrapper = () => {
      const [value, setValue] = React.useState<Item[]>(items);

      const handleChange = (ev: ChangeEvent) => {
        expect(ev).toMatchObject(eventMatcher);

        setValue(ev.component.value);
      };

      return (
        <Notifications
          value={value}
          onChange={handleChange}
          maxItems={3}
        />
      );
    };

    const { container } = render(<Wrapper />);

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(2);

    expect(container).toMatchSnapshot();

    act(() => {
      jest.advanceTimersByTime(100);
    });

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(1);

    expect(container).toMatchSnapshot();
  });

  it('should delete items by close icon click', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 0,
        id: 1,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'danger',
        color: 'danger',
        delay: 0,
        id: 2,
      },
    ];

    const eventMatcher = expect.objectContaining({
      component: expect.objectContaining({
        value: expect.any(Array),
        method: 'close-icon-click',
      }),
    });

    const Wrapper = () => {
      const [value, setValue] = React.useState<Item[]>(items);

      const handleChange = (ev: ChangeEvent) => {
        expect(ev).toMatchObject(eventMatcher);

        setValue(ev.component.value);
      };

      return (
        <Notifications
          value={value}
          onChange={handleChange}
          maxItems={3}
        />
      );
    };

    const { container } = render(<Wrapper />);

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(2);

    expect(container).toMatchSnapshot();

    act(() => {
      const icon = container.querySelector('.notifications-icon-close');

      expect(icon).toBeDefined();

      userEvent.click(icon as HTMLElement);
    });

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(1);

    act(() => {
      const icon = container.querySelector('.notifications-icon-close');

      expect(icon).toBeDefined();

      userEvent.click(icon as HTMLElement);
    });

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(0);

    expect(container).toMatchSnapshot();
  });
});

describe('Notifications add item', () => {
  it('should add new item', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 10,
        id: 1,
      },
    ];

    const item = {
      text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
      icon: 'danger',
      color: 'danger',
      delay: 10,
      id: 2,
    };

    const Wrapper = () => {
      const [value, setValue] = React.useState<Item[]>(items);

      const handleChange = (ev: ChangeEvent) => {
        setValue(ev.component.value);
      };

      const addNewItem = () => setValue([...value, item]);

      return (
        <>
          <Notifications
            value={value}
            onChange={handleChange}
            maxItems={3}
          />
          <Button onClick={() => addNewItem()} />
        </>
      );
    };

    const { container } = render(<Wrapper />);

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(1);

    act(() => {
      const button = container.querySelector('button');

      expect(button).toBeDefined();

      userEvent.click(button as HTMLElement);
    });

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(2);

    expect(container).toMatchSnapshot();
  });

  it('should add new items and delete more than maxItems', () => {
    const items = [
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 10,
        id: 1,
      },
      {
        text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
        icon: 'success',
        color: 'success',
        delay: 10,
        id: 3,
      },
    ];

    const item = {
      text: 'Сомнение определяет <a href="#">эмпирический катарсис</a>, не&nbsp;учитывая мнения авторитетов.',
      icon: 'danger',
      color: 'danger',
      delay: 10,
    };

    const Wrapper = () => {
      const [value, setValue] = React.useState<Item[]>(items);

      const handleChange = (ev: ChangeEvent) => {
        setValue(ev.component.value);
      };

      const addNewItem = () => {
        setValue([...value, { ...item, id: shortid.generate() }]);
      };

      return (
        <>
          <Notifications
            value={value}
            onChange={handleChange}
            maxItems={3}
          />
          <Button onClick={() => addNewItem()} />
        </>
      );
    };

    const { container } = render(<Wrapper />);

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(2);

    act(() => {
      const button = container.querySelector('button');

      expect(button).toBeDefined();

      userEvent.click(button as HTMLElement);
    });

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(3);

    act(() => {
      const button = container.querySelector('button');

      expect(button).toBeDefined();

      userEvent.click(button as HTMLElement);
    });

    expect(container.querySelectorAll('.notifications-item')).toHaveLength(3);

    act(() => {
      const button = container.querySelector('button');

      expect(button).toBeDefined();

      userEvent.click(button as HTMLElement);
    });
    // не более 3 элементов за раз
    expect(container.querySelectorAll('.notifications-item')).toHaveLength(3);

    expect(container).toMatchSnapshot();
  });
});
