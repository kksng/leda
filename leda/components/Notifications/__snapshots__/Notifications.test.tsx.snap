// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications SNAPSHOTS should render action button 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
        <button
          class="button-wrapper"
          type="button"
        >
          text
        </button>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
        <button
          class="button-wrapper"
          type="button"
        >
          text
        </button>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
</div>
`;

exports[`Notifications SNAPSHOTS should render basic usage 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
</div>
`;

exports[`Notifications SNAPSHOTS should render items < maxItems 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
</div>
`;

exports[`Notifications add item should add new item 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
  <button
    class="button-wrapper"
    type="button"
  />
</div>
`;

exports[`Notifications add item should add new items and delete more than maxItems 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
  <button
    class="button-wrapper"
    type="button"
  />
</div>
`;

exports[`Notifications handlers should delete items by close icon click 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
</div>
`;

exports[`Notifications handlers should delete items by close icon click 2`] = `<div />`;

exports[`Notifications handlers should not delete items with delay 0 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
</div>
`;

exports[`Notifications handlers should not delete items with delay 0 2`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
</div>
`;

exports[`Notifications handlers should trigger onChange close 1`] = `
<div>
  <div
    class="notifications-wrapper"
  >
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
    <div
      class="notifications-item"
    >
      <div
        class="notifications-icon"
      />
      <div
        class="notifications-content"
      >
        <div>
          Сомнение определяет 
          <a
            href="#"
          >
            эмпирический катарсис
          </a>
          , не учитывая мнения авторитетов.
        </div>
      </div>
      <div
        class="notifications-icon-close"
      />
    </div>
  </div>
</div>
`;

exports[`Notifications handlers should trigger onChange close 2`] = `<div />`;
